pkg/storage/permanent/settings.go

```diff
@@ -105,6 +109,31 @@ func (s *Settings) APIForVersion(version iotago.Version) iotago.API {
 	}
 }
 
+func (s *Settings) apiForVersion(version iotago.Version) iotago.API {
+	if a, exists := s.apiByVersion[version]; exists {
+		return a
+	}
+
+	protocolParams := s.protocolParameters(version)
+
+	if protocolParams == nil {
+		panic(fmt.Errorf("protocol parameters for version %d not found", version))
+	}
+
+	var a iotago.API
+	switch protocolParams.Version() {
+	case 3:
+		a = iotago.V3API(protocolParams)
+		s.apiMutex.Lock()
+		s.apiByVersion[3] = a
+		s.apiMutex.Unlock()
+
+		return a
+	default:
+		panic(ierrors.Errorf("unsupported protocol version %d", protocolParams.Version()))
+	}
```

Comment:
Maybe this could be refactored into a function `apiFromProtocolParameters` so you don't need to duplicate the code.

Reply:
```golang

func (s *Settings) apiForVersion(version iotago.Version) iotago.API {
	if api, exists := s.apiByVersion[version]; exists {
		return api
	}
	
	api := s.apiFromProtocolParameters(version)

	s.apiMutex.Lock()
	s.apiByVersion[version] = api
	s.apiMutex.Unlock()

	return api
}

func (s *Settings) APIForVersion(version iotago.Version) iotago.API {
	s.apiMutex.RLock()
	if api, exists := s.apiByVersion[version]; exists {
		s.apiMutex.RUnlock()
		return api
	}
	s.apiMutex.RUnlock()

	api := s.apiFromProtocolParameters(version)

	s.apiMutex.Lock()
	s.apiByVersion[version] = api
	s.apiMutex.Unlock()

	return api
}

func (s *Settings) apiFromProtocolParameters(version iotago.Version) iotago.API {
	protocolParams := s.protocolParameters(version)
	if protocolParams == nil {
		panic(fmt.Errorf("protocol parameters for version %d not found", version))
	}

	switch protocolParams.Version() {
	case 3:
		return iotago.V3API(protocolParams)
	default:
		panic(ierrors.Errorf("unsupported protocol version %d", protocolParams.Version()))
	}
}
```